pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -827,6 +834,119 @@ allChainsLabel:
 	return mutable.Listener
 }
 
+// TODO: abstract the common part, and reuse with buildGatewayListenerTLSContext
+func buildSidecarListenerTLSContext(tlsOpts *networking.Server_TLSOptions, sdsPath string, metadata *model.NodeMetadata) *auth.DownstreamTlsContext {
+	if tlsOpts == nil || (tlsOpts.Mode == networking.Server_TLSOptions_PASSTHROUGH) ||
+		(tlsOpts.Mode == networking.Server_TLSOptions_AUTO_PASSTHROUGH) {
+		return nil
+	}
+
+	tls := &auth.DownstreamTlsContext{
+		CommonTlsContext: &auth.CommonTlsContext{
+			AlpnProtocols: util.ALPNHttp,
+		},
+	}
+
+	if tlsOpts.Mode == networking.Server_TLSOptions_MUTUAL ||
+		tlsOpts.Mode == networking.Server_TLSOptions_ISTIO_MUTUAL {
+		tls.RequireClientCertificate = proto.BoolTrue
+	} else {
+		tls.RequireClientCertificate = proto.BoolFalse
+	}
+
+	if tlsOpts.Mode == networking.Server_TLSOptions_ISTIO_MUTUAL {
+		// global SDS must be enabled if the sds path is non-default, configure TLS with SDS
+		if sdsPath != "" {
```

Comment:
The following code seems a huge duplicate of the BuildInboundFilterChain() in pilot/pkg/security/authn/utils/utils.go, could you refactor to share the same code path?

Reply:
Yes, and this function is also mostly same as buildGatewayTlscontext.

Reply:
done